src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -548,20 +548,34 @@ public async Task<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request,
                     ParseStatusLine(await ReadNextResponseHeaderLineAsync().ConfigureAwait(false), response);
                 }
 
-                // If we sent an Expect: 100-continue header, and didn't receive a 100-continue. Handle the final response accordingly.
-                // Note that the developer may have added an Expect: 100-continue header even if there is no Content.
+                // Parse the response headers.
+                while (true)
+                {
+                    ArraySegment<byte> line = await ReadNextResponseHeaderLineAsync(foldedHeadersAllowed: true).ConfigureAwait(false);
+                    if (IsLineEmpty(line))
+                    {
+                        break;
+                    }
+                    ParseHeaderNameValue(this, line, response);
+                }
+
                 if (allowExpect100ToContinue != null)
                 {
+                    // If we sent an Expect: 100-continue header, and didn't receive a 100-continue. Handle the final response accordingly.
+                    // Note that the developer may have added an Expect: 100-continue header even if there is no Content.
                     if ((int)response.StatusCode >= 300 &&
                         request.Content != null &&
-                        (request.Content.Headers.ContentLength == null || request.Content.Headers.ContentLength.GetValueOrDefault() > Expect100ErrorSendThreshold))
+                        (request.Content.Headers.ContentLength == null || request.Content.Headers.ContentLength.GetValueOrDefault() > Expect100ErrorSendThreshold) &&
+                        !((int)response.StatusCode == 401 && AuthenticationHelper.IsSessionAuthenticationChallenge(response)))
```

Comment:
To simplify this expression, can we move the response.StatusCode check into the IsSessionAuthenticationChallenge helper?

Reply:
Also, it can be `response.StatusCode == HttpStatusCode.Unauthorized` rather than `(int)response.StatusCode == 401`.

Reply:
I like the suggestion. Code is updated now.